File: app/src/main/java/com/metahumanz/pacilread/reader/modern/dialog/ReaderOptionsDialogController.java
package com.metahumanz.pacilread.reader.modern.dialog;

import android.app.AlertDialog;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.TextView;

import com.metahumanz.pacilread.R;
import com.metahumanz.pacilread.reader.modern.ModernReaderActivity;
import com.metahumanz.pacilread.reader.modern.ReaderRuntime;
import com.metahumanz.pacilread.reader.modern.ReaderSessionState;
import com.metahumanz.pacilread.reader.modern.ReaderUiUtils;
import com.metahumanz.pacilread.reader.modern.config.ReaderOptionCatalog;
import com.metahumanz.pacilread.reader.modern.content.ReaderContentController;
import com.metahumanz.pacilread.reader.modern.paging.ReaderNavigationController;
import com.metahumanz.pacilread.reader.modern.ui.ReaderChromeController;
import com.metahumanz.pacilread.reader.modern.ui.ReaderStyleController;

public final class ReaderOptionsDialogController {
    private final ModernReaderActivity activity;
    private final ReaderRuntime runtime;
    private final ReaderSessionState state;
    private final ReaderDialogSupport dialogSupport;
    private final ReaderContentController content;
    private final ReaderNavigationController navigation;
    private final ReaderStyleController style;
    private final ReaderChromeController chrome;

    public ReaderOptionsDialogController(
            ModernReaderActivity activity,
            ReaderRuntime runtime,
            ReaderSessionState state,
            ReaderUiUtils ui,
            ReaderDialogSupport dialogSupport,
            ReaderContentController content,
            ReaderNavigationController navigation,
            ReaderStyleController style,
            ReaderChromeController chrome
    ) {
        this.activity = activity;
        this.runtime = runtime;
        this.state = state;
        this.dialogSupport = dialogSupport;
        this.content = content;
        this.navigation = navigation;
        this.style = style;
        this.chrome = chrome;
    }

    public void showReaderOptionsDialog() {
        View contentView = LayoutInflater.from(activity).inflate(R.layout.dialog_reader_options, null, false);
        OptionsDialogViews refs = OptionsDialogViews.bind(contentView);
        dialogSupport.applyTocStyleFullscreenInsets(contentView, refs.contentContainer);
        refs.titleInput.setText(state.book == null ? "" : state.book.title);
        refs.authorInput.setText(state.book == null ? "" : state.book.author);
        refs.showTitleCheck.setChecked(runtime.settingsStore.isChapterTitleVisible());
        refs.persistentActionsCheck.setChecked(runtime.settingsStore.isReaderMenuPersistentActionsEnabled());

        ArrayAdapter<String> flipAdapter = dialogSupport.buildSpinnerAdapter(new String[]{"覆盖", "平移", "仿真", "滚动", "无动画"});
        refs.flipSpinner.setAdapter(flipAdapter);
        refs.flipSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.FLIP_KEYS, runtime.settingsStore.getFlipMode(), 0), false);

        String[] speedKeys = new String[]{"fast", "medium", "slow"};
        ArrayAdapter<String> speedAdapter = dialogSupport.buildSpinnerAdapter(new String[]{"较快", "适中", "较慢"});
        refs.flipSpeedSpinner.setAdapter(speedAdapter);
        refs.flipSpeedSpinner.setSelection(ReaderOptionCatalog.indexOf(speedKeys, runtime.settingsStore.getFlipSpeed(), 1), false);

        final String[] sliderMode = new String[]{runtime.settingsStore.getReaderSliderMode()};
        AlertDialog dialog = new AlertDialog.Builder(activity).setView(contentView).create();
        dialogSupport.addAlignedCloseButton(contentView, R.id.options_title, refs.contentContainer, dialog);
        chrome.styleThemeButton(refs.sliderBookButton, "book".equals(sliderMode[0]));
        chrome.styleThemeButton(refs.sliderChapterButton, "chapter".equals(sliderMode[0]));

        ArrayAdapter<String> hudAdapter = dialogSupport.buildSpinnerAdapter(
                new String[]{"无", "书名", "章节名", "书名 / 章节名", "现在时间", "系统电量", "本章页数进度", "全书进度", "页数及进度", "时间及电量"}
        );
        refs.topLeftSpinner.setAdapter(hudAdapter);
        refs.topLeftSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, runtime.settingsStore.getHudTopLeft(), 0), false);
        refs.topCenterSpinner.setAdapter(hudAdapter);
        refs.topCenterSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, runtime.settingsStore.getHudTopCenter(), 0), false);
        refs.topRightSpinner.setAdapter(hudAdapter);
        refs.topRightSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, runtime.settingsStore.getHudTopRight(), 0), false);
        refs.bottomLeftSpinner.setAdapter(hudAdapter);
        refs.bottomLeftSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, runtime.settingsStore.getHudBottomLeft(), 0), false);
        refs.bottomCenterSpinner.setAdapter(hudAdapter);
        refs.bottomCenterSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, runtime.settingsStore.getHudBottomCenter(), 0), false);
        refs.bottomRightSpinner.setAdapter(hudAdapter);
        refs.bottomRightSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, runtime.settingsStore.getHudBottomRight(), 0), false);

        refs.hudTopMarginSeek.setProgress(runtime.settingsStore.getHudTopMarginDp());
        refs.hudBottomMarginSeek.setProgress(runtime.settingsStore.getHudBottomMarginDp());
        updateHudMarginLabels(refs);

        Runnable autoApply = () -> {
            String title = refs.titleInput.getText().toString().trim();
            String author = refs.authorInput.getText().toString().trim();
            if (title.isEmpty()) {
                title = "未命名书籍";
            }
            String finalTitle = title;
            String finalAuthor = author;
            int anchorOffset = content.currentCharOffset();
            boolean chapterTitleVisibilityChanged = runtime.settingsStore.isChapterTitleVisible() != refs.showTitleCheck.isChecked();
            if (state.book != null) {
                state.book.title = finalTitle;
                state.book.author = finalAuthor;
            }
            runtime.settingsStore.setFlipMode(ReaderOptionCatalog.FLIP_KEYS[refs.flipSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setFlipSpeed(speedKeys[refs.flipSpeedSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setReaderSliderMode(sliderMode[0]);
            runtime.settingsStore.setChapterTitleVisible(refs.showTitleCheck.isChecked());
            runtime.settingsStore.setReaderMenuPersistentActionsEnabled(refs.persistentActionsCheck.isChecked());
            runtime.settingsStore.setHudTopLeft(ReaderOptionCatalog.HUD_KEYS[refs.topLeftSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setHudTopCenter(ReaderOptionCatalog.HUD_KEYS[refs.topCenterSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setHudTopRight(ReaderOptionCatalog.HUD_KEYS[refs.topRightSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setHudBottomLeft(ReaderOptionCatalog.HUD_KEYS[refs.bottomLeftSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setHudBottomCenter(ReaderOptionCatalog.HUD_KEYS[refs.bottomCenterSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setHudBottomRight(ReaderOptionCatalog.HUD_KEYS[refs.bottomRightSpinner.getSelectedItemPosition()]);
            runtime.settingsStore.setHudTopMarginDp(refs.hudTopMarginSeek.getProgress());
            runtime.settingsStore.setHudBottomMarginDp(refs.hudBottomMarginSeek.getProgress());
            if (chapterTitleVisibilityChanged) {
                content.scheduleReflowAfterLayout(state.currentChapterIndex, anchorOffset);
            } else {
                chrome.updateReaderLayoutInsets();
                chrome.applyMenuLayoutMode();
                chrome.updateUiAfterPageChange();
            }
            runtime.executor.execute(() -> runtime.databaseHelper.updateBookInfo(state.bookId, finalTitle, finalAuthor));
        };

        TextWatcher textWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                autoApply.run();
            }
        };
        refs.titleInput.addTextChangedListener(textWatcher);
        refs.authorInput.addTextChangedListener(textWatcher);
        refs.showTitleCheck.setOnCheckedChangeListener((buttonView, isChecked) -> autoApply.run());
        refs.persistentActionsCheck.setOnCheckedChangeListener((buttonView, isChecked) -> autoApply.run());
        SeekBar.OnSeekBarChangeListener hudMarginSeekListener = new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                updateHudMarginLabels(refs);
                if (fromUser) {
                    autoApply.run();
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        };
        refs.hudTopMarginSeek.setOnSeekBarChangeListener(hudMarginSeekListener);
        refs.hudBottomMarginSeek.setOnSeekBarChangeListener(hudMarginSeekListener);

        android.widget.AdapterView.OnItemSelectedListener flipListener = new android.widget.AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(android.widget.AdapterView<?> parent, View view, int position, long id) {
                autoApply.run();
            }

            @Override
            public void onNothingSelected(android.widget.AdapterView<?> parent) {
            }
        };
        refs.flipSpinner.setOnItemSelectedListener(flipListener);
        refs.flipSpeedSpinner.setOnItemSelectedListener(flipListener);

        final Spinner[] allHudSpinners = {
                refs.topLeftSpinner, refs.topCenterSpinner, refs.topRightSpinner,
                refs.bottomLeftSpinner, refs.bottomCenterSpinner, refs.bottomRightSpinner
        };
        final boolean[] isAdjustingHudSpinners = new boolean[]{false};

        android.widget.AdapterView.OnItemSelectedListener hudListener = new android.widget.AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(android.widget.AdapterView<?> parent, View view, int position, long id) {
                if (position > 0 && !isAdjustingHudSpinners[0]) {
                    isAdjustingHudSpinners[0] = true;
                    for (Spinner spinner : allHudSpinners) {
                        if (spinner != parent && spinner.getSelectedItemPosition() == position) {
                            spinner.setSelection(0, false);
                        }
                    }
                    isAdjustingHudSpinners[0] = false;
                }
                autoApply.run();
            }

            @Override
            public void onNothingSelected(android.widget.AdapterView<?> parent) {
            }
        };
        refs.topLeftSpinner.setOnItemSelectedListener(hudListener);
        refs.topCenterSpinner.setOnItemSelectedListener(hudListener);
        refs.topRightSpinner.setOnItemSelectedListener(hudListener);
        refs.bottomLeftSpinner.setOnItemSelectedListener(hudListener);
        refs.bottomCenterSpinner.setOnItemSelectedListener(hudListener);
        refs.bottomRightSpinner.setOnItemSelectedListener(hudListener);

        refs.sliderBookButton.setOnClickListener(v -> {
            sliderMode[0] = "book";
            chrome.styleThemeButton(refs.sliderBookButton, true);
            chrome.styleThemeButton(refs.sliderChapterButton, false);
            autoApply.run();
        });
        refs.sliderChapterButton.setOnClickListener(v -> {
            sliderMode[0] = "chapter";
            chrome.styleThemeButton(refs.sliderBookButton, false);
            chrome.styleThemeButton(refs.sliderChapterButton, true);
            autoApply.run();
        });

        dialogSupport.showImmersiveFullscreenDialog(dialog, state.controlsVisible);
        contentView.requestApplyInsets();
    }

    private void updateHudMarginLabels(OptionsDialogViews refs) {
        refs.hudTopMarginValue.setText(refs.hudTopMarginSeek.getProgress() + " dp");
        refs.hudBottomMarginValue.setText(refs.hudBottomMarginSeek.getProgress() + " dp");
    }

    private static final class OptionsDialogViews {
        final EditText titleInput;
        final EditText authorInput;
        final View contentContainer;
        final CheckBox showTitleCheck;
        final CheckBox persistentActionsCheck;
        final Spinner flipSpinner;
        final Spinner flipSpeedSpinner;
        final Button sliderBookButton;
        final Button sliderChapterButton;
        final SeekBar hudTopMarginSeek;
        final SeekBar hudBottomMarginSeek;
        final TextView hudTopMarginValue;
        final TextView hudBottomMarginValue;
        final Spinner topLeftSpinner;
        final Spinner topCenterSpinner;
        final Spinner topRightSpinner;
        final Spinner bottomLeftSpinner;
        final Spinner bottomCenterSpinner;
        final Spinner bottomRightSpinner;

        private OptionsDialogViews(View root) {
            contentContainer = root.findViewById(R.id.options_content);
            titleInput = root.findViewById(R.id.options_input_title);
            authorInput = root.findViewById(R.id.options_input_author);
            showTitleCheck = root.findViewById(R.id.options_check_show_title);
            persistentActionsCheck = root.findViewById(R.id.options_check_persistent_actions);
            flipSpinner = root.findViewById(R.id.options_spinner_flip_mode);
            flipSpeedSpinner = root.findViewById(R.id.options_spinner_flip_speed);
            sliderBookButton = root.findViewById(R.id.options_button_slider_book);
            sliderChapterButton = root.findViewById(R.id.options_button_slider_chapter);
            hudTopMarginSeek = root.findViewById(R.id.options_seek_hud_top_margin);
            hudBottomMarginSeek = root.findViewById(R.id.options_seek_hud_bottom_margin);
            hudTopMarginValue = root.findViewById(R.id.options_text_hud_top_margin_value);
            hudBottomMarginValue = root.findViewById(R.id.options_text_hud_bottom_margin_value);
            topLeftSpinner = root.findViewById(R.id.options_spinner_hud_tl);
            topCenterSpinner = root.findViewById(R.id.options_spinner_hud_tc);
            topRightSpinner = root.findViewById(R.id.options_spinner_hud_tr);
            bottomLeftSpinner = root.findViewById(R.id.options_spinner_hud_bl);
            bottomCenterSpinner = root.findViewById(R.id.options_spinner_hud_bc);
            bottomRightSpinner = root.findViewById(R.id.options_spinner_hud_br);
        }

        static OptionsDialogViews bind(View root) {
            return new OptionsDialogViews(root);
        }
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/reader/modern/dialog/ReaderOptionsDialogController.kt
package com.metahumanz.pacilread.reader.modern.dialog

import android.app.AlertDialog
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.widget.AdapterView
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.Spinner
import android.widget.TextView
import com.metahumanz.pacilread.R
import com.metahumanz.pacilread.reader.modern.ModernReaderActivity
import com.metahumanz.pacilread.reader.modern.ReaderRuntime
import com.metahumanz.pacilread.reader.modern.ReaderSessionState
import com.metahumanz.pacilread.reader.modern.ReaderUiUtils
import com.metahumanz.pacilread.reader.modern.config.ReaderOptionCatalog
import com.metahumanz.pacilread.reader.modern.content.ReaderContentController
import com.metahumanz.pacilread.reader.modern.paging.ReaderNavigationController
import com.metahumanz.pacilread.reader.modern.ui.ReaderChromeController
import com.metahumanz.pacilread.reader.modern.ui.ReaderStyleController
import com.metahumanz.pacilread.stats.ReadingStatsUtils

class ReaderOptionsDialogController(
    private val activity: ModernReaderActivity,
    private val runtime: ReaderRuntime,
    private val state: ReaderSessionState,
    ui: ReaderUiUtils,
    private val dialogSupport: ReaderDialogSupport,
    private val content: ReaderContentController,
    private val navigation: ReaderNavigationController,
    private val style: ReaderStyleController,
    private val chrome: ReaderChromeController,
) {
    fun showReaderOptionsDialog() {
        val contentView = LayoutInflater.from(activity).inflate(R.layout.dialog_reader_options, null, false)
        val refs = OptionsDialogViews.bind(contentView)
        dialogSupport.applyTocStyleFullscreenInsets(contentView, refs.contentContainer)
        refs.titleInput.setText(state.book?.title.orEmpty())
        refs.authorInput.setText(state.book?.author.orEmpty())
        refs.showTitleCheck.isChecked = runtime.settingsStore.isChapterTitleVisible
        refs.persistentActionsCheck.isChecked = runtime.settingsStore.isReaderMenuPersistentActionsEnabled

        refs.flipSpinner.adapter = dialogSupport.buildSpinnerAdapter(arrayOf("覆盖", "平移", "仿真", "滚动", "无动画"))
        refs.flipSpinner.setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.FLIP_KEYS, runtime.settingsStore.flipMode, 0), false)
        val speedKeys = arrayOf("fast", "medium", "slow")
        refs.flipSpeedSpinner.adapter = dialogSupport.buildSpinnerAdapter(arrayOf("较快", "适中", "较慢"))
        refs.flipSpeedSpinner.setSelection(ReaderOptionCatalog.indexOf(speedKeys, runtime.settingsStore.flipSpeed, 1), false)

        var sliderMode = runtime.settingsStore.readerSliderMode
        var simulationTurnMode = runtime.settingsStore.simulationDoublePageTurnMode
        val dialog = AlertDialog.Builder(activity).setView(contentView).create()
        dialogSupport.addAlignedCloseButton(contentView, R.id.options_title, refs.contentContainer, dialog)
        chrome.styleThemeButton(refs.sliderBookButton, sliderMode == "book")
        chrome.styleThemeButton(refs.sliderChapterButton, sliderMode == "chapter")
        applySimulationDoublePageTurnModeButtons(refs, simulationTurnMode)
        updateSimulationDoublePageTurnModeVisibility(refs)

        val hudAdapter = dialogSupport.buildSpinnerAdapter(arrayOf(
            "无", "书名", "章节名", "书名 / 章节名", "现在时间", "系统电量", "本章页数进度", "全书进度", "页数及进度", "时间及电量",
        ))
        val hudSpinners = arrayOf(refs.topLeftSpinner, refs.topCenterSpinner, refs.topRightSpinner,
            refs.bottomLeftSpinner, refs.bottomCenterSpinner, refs.bottomRightSpinner)
        val hudValues = arrayOf(runtime.settingsStore.hudTopLeft, runtime.settingsStore.hudTopCenter,
            runtime.settingsStore.hudTopRight, runtime.settingsStore.hudBottomLeft,
            runtime.settingsStore.hudBottomCenter, runtime.settingsStore.hudBottomRight)
        for (i in hudSpinners.indices) {
            hudSpinners[i].adapter = hudAdapter
            hudSpinners[i].setSelection(ReaderOptionCatalog.indexOf(ReaderOptionCatalog.HUD_KEYS, hudValues[i], 0), false)
        }
        refs.hudTopMarginSeek.progress = runtime.settingsStore.hudTopMarginDp
        refs.hudBottomMarginSeek.progress = runtime.settingsStore.hudBottomMarginDp
        updateHudMarginLabels(refs)

        val autoApply = Runnable {
            val title = refs.titleInput.text.toString().trim().ifEmpty { "未命名书籍" }
            val author = refs.authorInput.text.toString().trim()
            val anchorOffset = content.currentCharOffset()
            val chapterTitleVisibilityChanged = runtime.settingsStore.isChapterTitleVisible != refs.showTitleCheck.isChecked
            val previousFlipMode = runtime.settingsStore.flipMode
            val previousFlipSpeed = runtime.settingsStore.flipSpeed
            val previousSimulationTurnMode = runtime.settingsStore.simulationDoublePageTurnMode
            val nextFlipMode = ReaderOptionCatalog.FLIP_KEYS[refs.flipSpinner.selectedItemPosition]
            val nextFlipSpeed = speedKeys[refs.flipSpeedSpinner.selectedItemPosition]
            val flipSettingsChanged = previousFlipMode != nextFlipMode || previousFlipSpeed != nextFlipSpeed ||
                previousSimulationTurnMode != simulationTurnMode
            state.book?.apply {
                this.title = title
                this.author = author
                readingStatsKey = ReadingStatsUtils.buildBookIdentity(title, author)
            }
            runtime.settingsStore.flipMode = nextFlipMode
            runtime.settingsStore.flipSpeed = nextFlipSpeed
            runtime.settingsStore.simulationDoublePageTurnMode = simulationTurnMode
            runtime.settingsStore.readerSliderMode = sliderMode
            runtime.settingsStore.isChapterTitleVisible = refs.showTitleCheck.isChecked
            runtime.settingsStore.isReaderMenuPersistentActionsEnabled = refs.persistentActionsCheck.isChecked
            runtime.settingsStore.hudTopLeft = ReaderOptionCatalog.HUD_KEYS[refs.topLeftSpinner.selectedItemPosition]
            runtime.settingsStore.hudTopCenter = ReaderOptionCatalog.HUD_KEYS[refs.topCenterSpinner.selectedItemPosition]
            runtime.settingsStore.hudTopRight = ReaderOptionCatalog.HUD_KEYS[refs.topRightSpinner.selectedItemPosition]
            runtime.settingsStore.hudBottomLeft = ReaderOptionCatalog.HUD_KEYS[refs.bottomLeftSpinner.selectedItemPosition]
            runtime.settingsStore.hudBottomCenter = ReaderOptionCatalog.HUD_KEYS[refs.bottomCenterSpinner.selectedItemPosition]
            runtime.settingsStore.hudBottomRight = ReaderOptionCatalog.HUD_KEYS[refs.bottomRightSpinner.selectedItemPosition]
            runtime.settingsStore.hudTopMarginDp = refs.hudTopMarginSeek.progress
            runtime.settingsStore.hudBottomMarginDp = refs.hudBottomMarginSeek.progress
            when {
                chapterTitleVisibilityChanged -> content.scheduleReflowAfterLayout(state.currentChapterIndex, anchorOffset)
                flipSettingsChanged -> {
                    style.applyReaderSettings()
                    chrome.applyMenuLayoutMode()
                    navigation.refreshPagingPresentationAfterSettingsChange()
                }
                else -> {
                    chrome.updateReaderLayoutInsets()
                    chrome.applyMenuLayoutMode()
                    chrome.updateUiAfterPageChange()
                }
            }
            runtime.safeExecute(Runnable { runtime.databaseHelper.updateBookInfo(state.bookId, title, author) }, "update reader book info")
        }

        val textWatcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) = Unit
            override fun afterTextChanged(s: Editable?) = autoApply.run()
        }
        refs.titleInput.addTextChangedListener(textWatcher)
        refs.authorInput.addTextChangedListener(textWatcher)
        refs.showTitleCheck.setOnCheckedChangeListener { _, _ -> autoApply.run() }
        refs.persistentActionsCheck.setOnCheckedChangeListener { _, _ -> autoApply.run() }
        val hudMarginSeekListener = object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                updateHudMarginLabels(refs)
                if (fromUser) autoApply.run()
            }
            override fun onStartTrackingTouch(seekBar: SeekBar?) = Unit
            override fun onStopTrackingTouch(seekBar: SeekBar?) = Unit
        }
        refs.hudTopMarginSeek.setOnSeekBarChangeListener(hudMarginSeekListener)
        refs.hudBottomMarginSeek.setOnSeekBarChangeListener(hudMarginSeekListener)

        val flipListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                updateSimulationDoublePageTurnModeVisibility(refs)
                autoApply.run()
            }
            override fun onNothingSelected(parent: AdapterView<*>?) = Unit
        }
        refs.flipSpinner.onItemSelectedListener = flipListener
        refs.flipSpeedSpinner.onItemSelectedListener = flipListener

        var isAdjustingHudSpinners = false
        val hudListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                if (position > 0 && !isAdjustingHudSpinners) {
                    isAdjustingHudSpinners = true
                    for (spinner in hudSpinners) if (spinner !== parent && spinner.selectedItemPosition == position) {
                        spinner.setSelection(0, false)
                    }
                    isAdjustingHudSpinners = false
                }
                autoApply.run()
            }
            override fun onNothingSelected(parent: AdapterView<*>?) = Unit
        }
        for (spinner in hudSpinners) spinner.onItemSelectedListener = hudListener

        refs.sliderBookButton.setOnClickListener {
            sliderMode = "book"
            chrome.styleThemeButton(refs.sliderBookButton, true)
            chrome.styleThemeButton(refs.sliderChapterButton, false)
            autoApply.run()
        }
        refs.sliderChapterButton.setOnClickListener {
            sliderMode = "chapter"
            chrome.styleThemeButton(refs.sliderBookButton, false)
            chrome.styleThemeButton(refs.sliderChapterButton, true)
            autoApply.run()
        }
        refs.simulationOuterPageButton.setOnClickListener {
            simulationTurnMode = "outerPage"
            applySimulationDoublePageTurnModeButtons(refs, simulationTurnMode)
            autoApply.run()
        }
        refs.simulationSpreadButton.setOnClickListener {
            simulationTurnMode = "spread"
            applySimulationDoublePageTurnModeButtons(refs, simulationTurnMode)
            autoApply.run()
        }
        dialogSupport.showImmersiveFullscreenDialog(dialog, state.controlsVisible)
        contentView.requestApplyInsets()
    }

    private fun updateSimulationDoublePageTurnModeVisibility(refs: OptionsDialogViews) {
        val visible = content.isDoublePageActive() &&
            ReaderOptionCatalog.FLIP_KEYS[refs.flipSpinner.selectedItemPosition] == "simulation"
        val visibility = if (visible) View.VISIBLE else View.GONE
        refs.simulationTurnModeLabel.visibility = visibility
        refs.simulationTurnModeLayout.visibility = visibility
    }

    private fun applySimulationDoublePageTurnModeButtons(refs: OptionsDialogViews, mode: String?) {
        val spread = mode == "spread"
        refs.simulationOuterPageButton.isSelected = !spread
        refs.simulationSpreadButton.isSelected = spread
        chrome.styleThemeButton(refs.simulationOuterPageButton, !spread)
        chrome.styleThemeButton(refs.simulationSpreadButton, spread)
    }

    private fun updateHudMarginLabels(refs: OptionsDialogViews) {
        refs.hudTopMarginValue.text = "${refs.hudTopMarginSeek.progress} dp"
        refs.hudBottomMarginValue.text = "${refs.hudBottomMarginSeek.progress} dp"
    }

    private class OptionsDialogViews private constructor(root: View) {
        val contentContainer: View = root.findViewById(R.id.options_content)
        val titleInput: EditText = root.findViewById(R.id.options_input_title)
        val authorInput: EditText = root.findViewById(R.id.options_input_author)
        val showTitleCheck: CheckBox = root.findViewById(R.id.options_check_show_title)
        val persistentActionsCheck: CheckBox = root.findViewById(R.id.options_check_persistent_actions)
        val flipSpinner: Spinner = root.findViewById(R.id.options_spinner_flip_mode)
        val flipSpeedSpinner: Spinner = root.findViewById(R.id.options_spinner_flip_speed)
        val simulationTurnModeLabel: TextView = root.findViewById(R.id.options_text_simulation_double_page_turn_mode_label)
        val simulationTurnModeLayout: LinearLayout = root.findViewById(R.id.options_layout_simulation_double_page_turn_mode)
        val simulationOuterPageButton: Button = root.findViewById(R.id.options_button_simulation_outer_page)
        val simulationSpreadButton: Button = root.findViewById(R.id.options_button_simulation_spread)
        val sliderBookButton: Button = root.findViewById(R.id.options_button_slider_book)
        val sliderChapterButton: Button = root.findViewById(R.id.options_button_slider_chapter)
        val hudTopMarginSeek: SeekBar = root.findViewById(R.id.options_seek_hud_top_margin)
        val hudBottomMarginSeek: SeekBar = root.findViewById(R.id.options_seek_hud_bottom_margin)
        val hudTopMarginValue: TextView = root.findViewById(R.id.options_text_hud_top_margin_value)
        val hudBottomMarginValue: TextView = root.findViewById(R.id.options_text_hud_bottom_margin_value)
        val topLeftSpinner: Spinner = root.findViewById(R.id.options_spinner_hud_tl)
        val topCenterSpinner: Spinner = root.findViewById(R.id.options_spinner_hud_tc)
        val topRightSpinner: Spinner = root.findViewById(R.id.options_spinner_hud_tr)
        val bottomLeftSpinner: Spinner = root.findViewById(R.id.options_spinner_hud_bl)
        val bottomCenterSpinner: Spinner = root.findViewById(R.id.options_spinner_hud_bc)
        val bottomRightSpinner: Spinner = root.findViewById(R.id.options_spinner_hud_br)

        companion object { fun bind(root: View) = OptionsDialogViews(root) }
    }
}
